File: caracal-node/src/delegation.ts
/**
 * Copyright (C) 2026 Garudex Labs. All Rights Reserved.
 * Caracal, a product of Garudex Labs
 *
 * SDK Delegation Operations — Graph-Based Model.
 */

import { SDKRequest } from './adapters/base';
import { ScopeContext } from './context';

export class DelegationOperations {
  constructor(private readonly scope: ScopeContext) {}

  private buildReq(method: string, path: string, body?: Record<string, unknown>, params?: Record<string, unknown>): SDKRequest {
    return { method, path, headers: { ...this.scope.scopeHeaders() }, body, params };
  }

  private async exec(req: SDKRequest): Promise<unknown> {
    const scoped = await this.scope._hooks.fireBeforeRequest(req, this.scope.toScopeRef());
    try {
      const res = await this.scope._adapter.send(scoped);
      this.scope._hooks.fireAfterResponse(res, this.scope.toScopeRef());
      return res.body;
    } catch (e) {
      this.scope._hooks.fireError(e as Error);
      throw e;
    }
  }

  /** Create a delegation edge from source mandate to target subject. */
  async create(options: {
    sourceMandateId: string;
    childSubjectId: string;
    resourceScope: string[];
    actionScope: string[];
    validitySeconds: number;
    contextTags?: string[];
    metadata?: Record<string, unknown>;
  }): Promise<unknown> {
    const body: Record<string, unknown> = {
      source_mandate_id: options.sourceMandateId,
      child_subject_id: options.childSubjectId,
      resource_scope: options.resourceScope,
      action_scope: options.actionScope,
      validity_seconds: options.validitySeconds,
    };
    if (options.contextTags) body.context_tags = options.contextTags;
    if (options.metadata) body.metadata = options.metadata;
    return this.exec(this.buildReq('POST', '/delegations', body));
  }

  /** Generate a delegation token from source agent to target agent. */
  async getToken(options: {
    sourceAgentId: string;
    targetAgentId: string;
    expirationSeconds?: number;
    allowedOperations?: string[];
  }): Promise<unknown> {
    const body: Record<string, unknown> = {
      source_agent_id: options.sourceAgentId,
      target_agent_id: options.targetAgentId,
      expiration_seconds: options.expirationSeconds ?? 86400,
    };
    if (options.allowedOperations) body.allowed_operations = options.allowedOperations;
    return this.exec(this.buildReq('POST', '/delegations/token', body));
  }

  /** Peer-to-peer delegation between principals of the same type. */
  async peerDelegate(options: {
    sourceMandateId: string;
    targetSubjectId: string;
    resourceScope: string[];
    actionScope: string[];
    validitySeconds: number;
    contextTags?: string[];
  }): Promise<unknown> {
    const body: Record<string, unknown> = {
      source_mandate_id: options.sourceMandateId,
      target_subject_id: options.targetSubjectId,
      resource_scope: options.resourceScope,
      action_scope: options.actionScope,
      validity_seconds: options.validitySeconds,
    };
    if (options.contextTags) body.context_tags = options.contextTags;
    return this.exec(this.buildReq('POST', '/delegations/peer', body));
  }

  /** Get delegation graph for an agent (all connected edges). */
  async getGraph(agentId: string): Promise<unknown> {
    return this.exec(this.buildReq('GET', `/delegations/graph/${agentId}`));
  }

  /** Revoke a specific delegation edge. */
  async revokeEdge(edgeId: string, cascade: boolean = true): Promise<unknown> {
    return this.exec(this.buildReq('DELETE', `/delegations/edges/${edgeId}`, undefined, { cascade }));
  }
}

File: caracal/sdk/delegation.py
"""
Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
Caracal, a product of Garudex Labs

SDK Delegation Operations.

Provides delegation token management within a scoped context.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Dict, List, Optional

from caracal.logging_config import get_logger
from caracal.sdk.adapters.base import SDKRequest

if TYPE_CHECKING:
    from caracal.sdk.context import ScopeContext

logger = get_logger(__name__)


class DelegationOperations:
    """Delegation chain and token management within a scoped context.

    All methods inject scope headers and fire lifecycle hooks.
    """

    def __init__(self, scope: ScopeContext) -> None:
        self._scope = scope

    def _build_request(
        self,
        method: str,
        path: str,
        body: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
    ) -> SDKRequest:
        headers = dict(self._scope.scope_headers())
        return SDKRequest(
            method=method, path=path, headers=headers, body=body, params=params
        )

    async def _execute(self, request: SDKRequest) -> Any:
        scope_ref = self._scope.to_scope_ref()
        request = self._scope._hooks.fire_before_request(request, scope_ref)
        try:
            response = await self._scope._adapter.send(request)
            self._scope._hooks.fire_after_response(response, scope_ref)
            return response.body
        except Exception as exc:
            self._scope._hooks.fire_error(exc)
            raise

    # -- Public API --------------------------------------------------------

    async def create(
        self,
        parent_mandate_id: str,
        child_subject_id: str,
        resource_scope: List[str],
        action_scope: List[str],
        validity_seconds: int,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Create a delegated mandate from a parent mandate."""
        body: Dict[str, Any] = {
            "parent_mandate_id": parent_mandate_id,
            "child_subject_id": child_subject_id,
            "resource_scope": resource_scope,
            "action_scope": action_scope,
            "validity_seconds": validity_seconds,
        }
        if metadata:
            body["metadata"] = metadata
        req = self._build_request("POST", "/delegations", body=body)
        return await self._execute(req)

    async def get_token(
        self,
        parent_agent_id: str,
        child_agent_id: str,
        expiration_seconds: int = 86400,
        allowed_operations: Optional[List[str]] = None,
    ) -> Dict[str, Any]:
        """Generate a delegation token for a child agent."""
        body: Dict[str, Any] = {
            "parent_agent_id": parent_agent_id,
            "child_agent_id": child_agent_id,
            "expiration_seconds": expiration_seconds,
        }
        if allowed_operations:
            body["allowed_operations"] = allowed_operations
        req = self._build_request("POST", "/delegations/token", body=body)
        return await self._execute(req)

    async def get_chain(self, agent_id: str) -> List[Dict[str, Any]]:
        """Get the delegation chain for an agent."""
        req = self._build_request("GET", f"/delegations/chain/{agent_id}")
        return await self._execute(req)
